File: Leetcode 2501-2600/2591/2591.py
# Leetcode 2591: Distribute Money to Maximum Children
# https://leetcode.com/problems/distribute-money-to-maximum-children/
# Solved on 25th of October, 2025
class Solution:
    def distMoney(self, money: int, children: int) -> int:
        """
        Distributes money to children such that as many children as possible receive exactly $8.

        Args:
            money (int): The total amount of money to distribute.
            children (int): The number of children.
        Returns:
            int: The maximum number of children that can receive exactly $8.
        """
        if money < children:
            return -1

        money -= children

        countEight = money // 7
        remainder = money % 7

        if countEight > children:
            return children - 1

        if countEight == children:
            if remainder == 0:
                return children
            else:
                return children - 1

        childrenLeft = children - countEight

        if childrenLeft == 1 and remainder == 3:
            return countEight - 1

        return countEight

File: Leetcode 2501-2600/2591/2591.java
// Leetcode 2591: Distribute Money to Maximum Children
// https://leetcode.com/problems/distribute-money-to-maximum-children/
// Solved on 25th of October, 2025
class Solution{
    /**
     * Distributes `money` among `children` such that each child receives at least 1 dollar,
     * and as many children as possible receive exactly 8 dollars.
     * @param money The total amount of money to distribute.
     * @param children The number of children.
     * @return The maximum number of children who can receive exactly 8 dollars. Returns -1 if it's impossible to give each child at least 1 dollar.
     */
    public int distMoney(int money, int children) {
        if (money < children) {
            return -1;
        }

        if (money > children * 8) {
            return children - 1;
        }

        if (money == children * 8) {
            return children;
        }

        int moneyAfterOnes = money - children;
        int countEight = moneyAfterOnes / 7;
        int moneyRemaining = moneyAfterOnes % 7;
        int childrenRemnaining = children - countEight;

        if (moneyRemaining == 3 && childrenRemnaining == 1) {
            return countEight - 1;
        }

        return countEight;
    }
}